packages/perpetual/src/fundingRateProposer.js

```diff
@@ -0,0 +1,223 @@
+const { getAbi } = require("@uma/core");
+
+class FundingRateProposer {
+  /**
+   * @notice Constructs new Perpetual FundingRate Proposer bot.
+   * @param {Object} logger Module used to send logs.
+   * @param {Object} perpetualFactoryClient Module used to query for live perpetual contracts
+   * @param {Object} gasEstimator Module used to estimate optimal gas price with which to send txns.
+   * @param {String} account Ethereum account from which to send txns.
+   * @param {Object} commonPriceFeedConfig Default configuration to construct all price feed objects.
+   * @param {Object} [perpetualProposerConfig] Contains fields with which constructor will attempt to override defaults.
+   */
+  constructor({
+    logger,
+    perpetualFactoryClient,
+    gasEstimator,
+    account,
+    commonPriceFeedConfig,
+    perpetualProposerConfig
+  }) {
+    this.logger = logger;
+    this.account = account;
+    this.perpetualFactoryClient = perpetualFactoryClient;
+    this.web3 = this.perpetualFactoryClient.web3;
+    this.commonPriceFeedConfig = commonPriceFeedConfig;
+
+    this.createPerpetualContract = perpAddress => {
+      return new this.web3.eth.Contract(getAbi("Perpetual"), perpAddress);
+    };
+
+    // Gas Estimator to calculate the current Fast gas rate.
+    this.gasEstimator = gasEstimator;
+
+    // Cached mapping of identifiers to pricefeed classes
+    this.priceFeedCache = {};
+
+    // Cached perpetual contracts
+    this.contractCache = {};
+
+    // Helper functions from web3.
+    this.BN = this.web3.utils.BN;
+    this.toBN = this.web3.utils.toBN;
+    this.toWei = this.web3.utils.toWei;
+    this.fromWei = this.web3.utils.fromWei;
+    this.utf8ToHex = this.web3.utils.utf8ToHex;
+    this.hexToUtf8 = this.web3.utils.hexToUtf8;
+
+    // Default config settings. Bot deployer can override these settings by passing in new
+    // values via the `perpetualProposerConfig` input object. The `isValid` property is a function that should be called
+    // before resetting any config settings. `isValid` must return a Boolean.
+    const defaultConfig = {
+      fundingRateErrorPercent: {
+        //   "fundingRateErrorPercent":0.05 ->  Current funding rates (as stored in the OptimisticOracle)
+        //                                      that do not equal the bot's queried funding rate
+        //                                      within this error % will be requested to be updated,
+        //                                      and proposed to.
+        //                                      e.g. 0.05 implies 5% margin of error.
+        value: 0.05,
+        isValid: x => {
+          return x >= 0 && x < 1;
+        }
+      }
+    };
+
+    // Validate and set config settings to class state.
+    const configWithDefaults = createObjectFromDefaultProps(perpetualProposerConfig, defaultConfig);
+    Object.assign(this, configWithDefaults);
+  }
+
+  async update() {
```

Comment:
do you not need to update the price feeds within `this.priceFeedCache`?

Reply:
I see you do this in line 129 per feed. would we not want to do them all at once in this update method?

Reply:
This is tough. I think the code would be cleaner if the `update()` for all pricefeeds was called in this `update()` method, but I left them for the `_updateFUndingRate()` method so that the pricefeeds could be updated as close as possible to their usage, to minimize gaps between when prices are fetched and whent hey are compared against on chain prices.